inout.go

```diff
@@ -22,18 +22,216 @@ package fx
+// provide a forward-compatible API: since adding fields to a struct is
+// backward-compatible, modules can then add optional dependencies in minor
+// releases.
+//
+// Parameter Structs
+//
+// Fx constructors declare their dependencies as function parameters. This can
+// quickly become unreadable if the constructor has a lot of dependencies.
+//
+//   func NewHandler(users *UserGateway, comments *CommentGateway, posts *PostGateway, votes *VoteGateway, authz *AuthZGateway) *Handler {
+//     // ...
+//   }
+//
+// To improve the readability of constructors like this, create a struct that
+// lists all the dependencies as fields and change the function to accept that
+// struct instead. The new struct is called a parameter struct.
+//
+// Fx has first class support for parameter structs: any struct embedding
+// fx.In gets treated as a parameter struct, so the individual fields in the
+// struct are supplied via dependency injection. Using a parameter struct, we
+// can make the constructor above much more readable:
+//
+//   type HandlerParams struct {
+//     fx.In
+//
+//     Users    *UserGateway
+//     Comments *CommentGateway
+//     Posts    *PostGateway
+//     Votes    *VoteGateway
+//     AuthZ    *AuthZGateway
+//   }
+//
+//   func NewHandler(p HandlerParams) *Handler {
+//     // ...
+//   }
+//
+// Though it's rarely a good idea, constructors can receive any combination of
+// parameter structs and parameters.
+//
+//   func NewHandler(p HandlerParams, l *log.Logger) *Handler {
+//     // ...
+//   }
+//
+// Optional Dependencies
+//
+// Constructors often have soft dependencies on some types: if those types are
+// missing, they can operate in a degraded state. Fx supports optional
+// dependencies via the `optional:"true"` tag to fields on parameter structs.
+//
+//   type UserGatewayParams struct {
+//     fx.In
+//
+//     Conn  *sql.DB
+//     Cache *redis.Client `optional:"true"`
+//   }
+//
+// If an optional field isn't available in the container, the constructor
+// receives the field's zero value.
+//
+//   func NewUserGateway(p UserGatewayParams, log *log.Logger) (*UserGateway, error) {
+//     if p.Cache != nil {
+//       log.Print("Logging disabled")
+//     }
+//     // ...
+//   }
+//
+// Constructors that declare optional dependencies MUST gracefully handle
+// situations in which those dependencies are absent.
+//
+// The optional tag also allows adding new dependencies without breaking
+// existing consumers of the constructor.
+//
+// Named Values
+//
+// Some use cases call for multiple values of the same type. Fx will allow
+// multiple values of the same type in an application's dependency injection
+// container via `name:".."` tags parameter and result structs.
+//
+// A constructor that produces a result struct can tag any field with
+// `name:".."` to have the corresponding value added to the graph under the
+// specified name.
+//
+//   type ConnectionResult struct {
+//     fx.Out
```

Comment:
The docs for fx.In talking about result structs and fx.Out before that is ever mentioned doesn't seem right. 

Perhaps advanced usage should be documented in the package docs like dig, or this should mention "see [..] for more information on how to produce named values from constructors".

Reply:
Makes sense. Discussed offline - plan it to separate the producer side (e.g., named value, value group) docs from the consume docs.